server/src/main/scala/services/graphstages/ArrivalSplitsGraphStage.scala

```diff
@@ -0,0 +1,307 @@
+          val pcpTime = SDate(updatedFlight.PcpTime).toLocalDateTimeString()
+          log.debug(s"Ignoring arrival with PCP time ($pcpTime) beyond latest crunch minute (${latestCrunchMinute.toLocalDateTimeString()})")
+          (updates, additions, flightsSoFar)
+        case ((updates, additions, flightsSoFar), updatedFlight) =>
+          flightsSoFar.get(updatedFlight.uniqueId) match {
+            case None =>
+              val ths = splitsCalculator.terminalAndHistoricSplits(updatedFlight)
+              val newFlightWithSplits = ApiFlightWithSplits(updatedFlight, ths, Option(SDate.now().millisSinceEpoch))
+              val newFlightWithAvailableSplits = addApiSplitsIfAvailable(newFlightWithSplits)
+              (updates, additions + 1, flightsSoFar.updated(updatedFlight.uniqueId, newFlightWithAvailableSplits))
+
+            case Some(existingFlight) if existingFlight.apiFlight != updatedFlight =>
+              (updates + 1, additions, flightsSoFar.updated(updatedFlight.uniqueId, existingFlight.copy(apiFlight = updatedFlight)))
+            case _ =>
+              (updates, additions, flightsSoFar)
+          }
+      }
+
+      log.info(s"${updatedFlights.size} flights after updates. $updatesCount updates & $additionsCount additions")
+
+      val minusExpired = purgeExpiredArrivals(updatedFlights)
+
+      val uniqueFlights = groupFlightsByCodeShares(minusExpired.values.toSeq)
+        .map { case (fws, _) => (fws.apiFlight.uniqueId, fws) }
+        .toMap
+
+      log.info(s"${uniqueFlights.size} flights after removing expired and accounting for codeshares")
+
+      uniqueFlights
+    }
+
+    def addApiSplitsIfAvailable(newFlightWithSplits: ApiFlightWithSplits): ApiFlightWithSplits = {
+      val arrival = newFlightWithSplits.apiFlight
+      val vmIdx = s"${Crunch.flightVoyageNumberPadded(arrival)}-${arrival.Scheduled}"
+
+      val newFlightWithAvailableSplits = manifestsBuffer.get(vmIdx) match {
+        case None => newFlightWithSplits
+        case Some(vm) =>
+          manifestsBuffer = manifestsBuffer.filterNot { case (idx, _) => idx == vmIdx }
+          log.debug(s"Found buffered manifest to apply to new flight, and removed from buffer")
+          removeManifestsOlderThan(twoDaysAgo)
+          updateFlightWithManifests(vm, newFlightWithSplits)
+      }
+      newFlightWithAvailableSplits
+    }
+
+    def removeManifestsOlderThan(thresholdMillis: MillisSinceEpoch): Unit = {
+      manifestsBuffer = manifestsBuffer.filter {
+        case (_, vmsInBuffer) =>
+          val vmsToKeep = vmsInBuffer.filter(vm => isNewerThan(thresholdMillis, vm))
+          vmsToKeep match {
+            case vms if vms.nonEmpty => true
+            case _ => false
+          }
+      }
+    }
+
+    def updateFlightsWithManifests(manifests: Set[VoyageManifest], flightsById: Map[Int, ApiFlightWithSplits]): Map[Int, ApiFlightWithSplits] = {
```

Comment:
This could also be broken into a matching function and an update function possibly.